src/drivers/indexeddb.js

```diff
@@ -181,27 +182,164 @@
             }
         }
 
+        // Initialize a singleton container for all running localForages.
+        if (!dbContexts) {
+            dbContexts = {};
+        }
+
+        // Get the current context of the database;
+        var dbContext = dbContexts[dbInfo.name];
+
+        // ...or create a new context.
+        if (!dbContext) {
+            dbContext = {
+                // Running localForages sharing a database.
+                forages: [],
+                // Shared database.
+                db: null
+            };
+            // Register the new context in the global container.
+            dbContexts[dbInfo.name] = dbContext;
+        }
+
+        // Register itself as a running localForage in the current context.
+        dbContext.forages.push(this);
+
+        // Create an array of readiness of the related localForages.
+        var ready = [];
+        for (var j = 0; j < dbContext.forages.length; j++) {
+            var forage = dbContext.forages[j];
+            if (forage !== this) { // Don't wait for itself...
+                ready.push(forage.ready());
+            }
+        }
+
+        // Take a snapshot of the related localForages.
+        var forages = dbContext.forages.slice(0);
+
+        // Initialize the connection process only when
+        // all the related localForages are ready.
+        return Promise.all(ready).then(function() {
+            dbInfo.db = dbContext.db;
+            // Get the connection or open a new one without upgrade.
+            return _getOriginalConnection(dbInfo);
+        }).then(function(db) {
+            dbInfo.db = db;
+            if (_isUpgradeNeeded(dbInfo, self._defaultConfig.version)) {
+                // Reopen the database for upgrading.
+                return _getUpgradedConnection(dbInfo);
+            }
+            return db;
+        }).then(function(db) {
+            dbInfo.db = dbContext.db = db;
+            self._dbInfo = dbInfo;
+            // Share the final connection amongst related localForages.
+            for (var k in forages) {
+                var forage = forages[k];
+                if (forage !== self) { // Self is already up-to-date.
+                    forage._dbInfo.db = dbInfo.db;
+                    forage._dbInfo.version = dbInfo.db;
+                }
+            }
+        });
```

Comment:
Cool, I can be patient. I’ll wait for a bit and check back in when it’s figured out :-)

- tofumatt

On 27 September 2015 at 20:26:39, Thodoris Greasidis (notifications@github.com) wrote:

In src/drivers/indexeddb.js:

> +                // Reopen the database for upgrading.
> +                return _getUpgradedConnection(dbInfo);
> +            }
> +            return db;
> +        }).then(function(db) {
> +            dbInfo.db = dbContext.db = db;
> +            self._dbInfo = dbInfo;
> +            // Share the final connection amongst related localForages.
> +            for (var k in forages) {
> +                var forage = forages[k];
> +                if (forage !== self) { // Self is already up-to-date.
> +                    forage._dbInfo.db = dbInfo.db;
> +                    forage._dbInfo.version = dbInfo.db;
> +                }
> +            }
> +        });
Now you made me hesitant about the completeness and correctness of my suggested fix... This is a way too small test case to leave for an other PR. On the other hand I feel ok with the current state of this PR. How about giving it an extra day or two?
On Sun, Sep 27, 2015, 22:16 Matthew Riley MacPherson < notifications@github.com> wrote: In src/drivers/indexeddb.js <https://github.com/mozilla/localForage/pull/452#discussion_r40508172>: > + // Reopen the database for upgrading. > + return _getUpgradedConnection(dbInfo); > + } > + return db; > + }).then(function(db) { > + dbInfo.db = dbContext.db = db; > + self._dbInfo = dbInfo; > + // Share the final connection amongst related localForages. > + for (var k in forages) { > + var forage = forages[k]; > + if (forage !== self) { // Self is already up-to-date. > + forage._dbInfo.db = dbInfo.db; > + forage._dbInfo.version = dbInfo.db; > + } > + } > + }); Should we just add in the catch and then merge this into master and address this use case later if needed then? I'd be okay with that... this patch is awesome. — Reply to this email directly or view it on GitHub <https://github.com/mozilla/localForage/pull/452/files#r40508172>.
-- *Thodoris Greasidis* Computer, Networking & Communications Engineer
—
Reply to this email directly or view it on GitHub.